calicoctl/resourcemgr/resourcemgr.go

```diff
@@ -351,46 +358,108 @@ func (rh resourceHelper) GetObjectType() reflect.Type {
 // Apply is an un-typed method to apply (create or update) a resource.  This calls directly
 // through to the resource helper specific Apply method which will map the untyped call to
 // the typed interface on the client.
-func (rh resourceHelper) Apply(client client.Interface, resource ResourceObject) (ResourceObject, error) {
-	return rh.apply(client, resource)
+func (rh resourceHelper) Apply(client client.Interface, resource ResourceObject, ns string) (runtime.Object, error) {
+	// Try to create the resource first.
+	ro, err := rh.Create(client, resource, ns)
+
+	switch err.(type) {
+	// Try updating if the resource already exists.
+	case cerrors.ErrorResourceAlreadyExists:
+		return rh.Update(client, resource, ns)
+
+	// For any other errors, return the error
+	default:
+		return ro, err
+	}
 }
 
 // Create is an un-typed method to create a new resource.  This calls directly
 // through to the resource helper specific Create method which will map the untyped call to
 // the typed interface on the client.
-func (rh resourceHelper) Create(client client.Interface, resource ResourceObject) (ResourceObject, error) {
-	return rh.create(client, resource)
+func (rh resourceHelper) Create(client client.Interface, resource ResourceObject, ns string) (runtime.Object, error) {
+	return rh.create(client, resource, ns)
 }
 
 // Update is an un-typed method to update an existing resource.  This calls directly
 // through to the resource helper specific Update method which will map the untyped call to
 // the typed interface on the client.
-func (rh resourceHelper) Update(client client.Interface, resource ResourceObject) (ResourceObject, error) {
-	return rh.update(client, resource)
+func (rh resourceHelper) Update(client client.Interface, resource ResourceObject, ns string) (runtime.Object, error) {
+	for i := 0; i < 5; i++ {
+		ro, err := rh.GetOrList(client, resource, ns, false)
+		if err != nil {
+			return ro, err
+		}
+		if resource.GetObjectMeta().GetName() != "" {
+			resource.GetObjectMeta().SetResourceVersion(ro.(ResourceObject).GetObjectMeta().GetResourceVersion())
+		} else {
+			resource.GetObjectMeta().SetResourceVersion(ro.(ResourceListObject).GetListMeta().GetResourceVersion())
+		}
+
+		ru, err := rh.update(client, resource, ns)
+		if err != nil {
+			switch err.(type) {
+			// Wait for a second and try again if there was a conflict during the resource update.
+			case cerrors.ErrorResourceUpdateConflict:
+				log.Warnf("Error updating the resource %s: %s. Retrying.", resource.GetObjectMeta().GetName(), err)
+				time.Sleep(1 * time.Second)
+				continue
+
+			// For any other errors, return the error
+			default:
+				return ru, err
+			}
+		}
+
+		return ru, err
+	}
+	return nil, errors.New("failed to update the resource after 5 retries")
```

Comment:
maybe just return the last error, or just return that there was an update conflict.